W-20600995 mcp tool mapping gr (#964)
mulesoft/docs-gateway: gateway/1.11/modules/ROOT/pages/policies-included-mcp-tool-mapping.adoc

= MCP Tool Mapping Policy
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:keywords: api gateway, flex gateway, gateway, policy, mcp, tool mapping

[width="100%", cols="5,15"]
|===
>s| Policy Name | MCP Tool Mapping
>s|Summary      | Renames MCP tool names in requests and responses to provide flexible tool naming
>s|Category | MCP
>s|First Flex Gateway version available | v1.11.0
>s| Returned Status Codes | No return codes exist for this policy.
|===

include::partial$mcp-version-support.adoc[]

== Summary

The MCP Tool Mapping policy intercepts MCP protocol traffic and applies configurable name transformations to tools. 

> This policy enables you to:

Reviewer: ```suggestion
With this policy, you can:
```

* Rename tools for readability
* Rename tools for consistency across different MCP servers
* Add prefixes or suffixes to tool names for namespacing

> * Create Pattern-based transformations by using regex

Reviewer: ```suggestion
* Create pattern-based transformations by using regex
```

* Create custom descriptions for mapped tools

The MCP Tool Mapping policy supports both outbound mapping (transforming tool names in `tools/list` responses from backend to client) and inbound mapping (reverse-transforming tool names in `tools/call` requests from client to backend). The policy automatically maintains bidirectional mapping, ensuring that tool names are correctly transformed in both directions to maintain consistency between client and backend.

The policy works with both MCP transport types:

* Standard JSON over HTTP: Transforms tool names in JSON-RPC messages
* Server-Sent Events (SSE): Transforms tool names in SSE streaming responses

== Configuring Policy Parameters

include::partial$policy-title-headers.adoc[tag=configFileTitleOnly]

The MCP Tool Mapping policy isn't supported in Local Mode.

include::partial$policy-title-headers.adoc[tag=ui]

[%header%autowidth.spread,cols="a,a"]
|===
| Element | Description
| Tool Name Mappings | Defines the tool name mappings. Each mapping specifies how to transform tool names between the client and backend.

> | Source Tool Name | The original tool name from the MCP server or a regex pattern for pattern-based matching.

Reviewer: Here, you've added the period even though it's not a complete sentence, but all the others are.

| Mapped Tool Name | The new tool name to use in responses. For regex mappings, use `$1`, `$2`, `$3`, and so on to reference capture groups.
| Mapped Tool Description | Optional new description for the tool. If not specified, the original description is used. New descriptions aren't supported for regex mappings.
| Mapping Type | Determines how the tool mapping is processed. Supported values are *Literal* for exact string matching and *Regex* for pattern matching with capture groups.
| Log Mappings | Logs tool names and mappings for debugging if enabled. For example:

----
Mapped outbound tool (literal): get_me -> ss_get_me
Passthrough outbound tool (no mapping): get_release_by_tag
----
|===

== Mapping Types

> 

Reviewer: Can you add some context here? For Agentforce chunking, we need text for all sections.

Reviewer: Alternatively, flatten the hierarchy. .

=== Literal Mapping

Literal mappings perform exact string matching for specific tool names. For example, given the following parameters:

* *Source Tool Name*: `get_weather`
* *Mapped Tool Name*: `weather_lookup`
* *Mapped Tool Description*: `Get current weather conditions`
* *Mapping Type*: *Literal*

The policy transforms the tool name in the `tools/list` response and `tools/call` request:

* `tools/list` response: `get_weather` to `weather_lookup`
* `tools/call` request: `weather_lookup` to `get_weather` (automatic reverse mapping)

=== Regex Mapping

> Regex mappings use pattern-based matching with capture group replacementFor example, given the following parameters:

Reviewer: ```suggestion
Regex mappings use pattern-based matching with capture group replacement. For example, given the following parameters:
```

* *Source Tool Name*: `get_(.+)`
* *Mapped Tool Name*: `fetch_$1`
* *Mapping Type*: *Regex*

The policy transforms the tool name in the `tools/list` response:

* `get_weather` to `fetch_weather`
* `get_user` to `fetch_user`
* `delete_item` remains unchanged (doesn't match pattern)

== Use Cases

> 

Reviewer: Add context here - or flatten the hierarchy.

=== Namespace Isolation

Add prefixes to avoid tool name conflicts between multiple MCP servers. For example, this configuration adds the prefix `server1_` to all tool names:

* *Source Tool Name*: `(.+)`
* *Mapped Tool Name*: `server1_$1`
* *Mapping Type*: `regex`

=== Consistent Naming

Standardize tool names across different servers. For example, this configuration standardizes the name of `fetch` and `retrieve` functions from different servers to `get` functions:

* Mapping 1:
** *Source Tool Name*: `fetchWeather`
** *Mapped Tool Name*: `get_weather`
** *Mapping Type*: `literal`

* Mapping 2:
** *Source Tool Name*: `retrieveUser`
** *Mapped Tool Name*: `get_user`
** *Mapping Type*: `literal`

== Bidirectional Mapping

> 

Reviewer: Context or flatten hierarchy

> === Outbound Mapping `tools/list` response

Reviewer: Don't use formatting in headings.
